### PR TITLE
pod issue investigator

### PR DESCRIPTION
**Pull request by Avi-Robusta**
Main Changes:

1) Refactored code that generates crashloop_backoff finding blocks to their own file
   - src/robusta/core/playbooks/pod_utils/crashloop_utils.py is from playbooks/robusta_playbooks/restart_loop_reporter.py

2) Refactored code that generates imagepull_backoff finding blocks to their own file
   - src/robusta/core/playbooks/pod_utils/imagepull_utils.py is from playbooks/robusta_playbooks/image_pull_backoff_enricher.py

3) Added enricher `pod_issue_investigator` that investigates if a pod is
- crashlooping (or crashed recently)
- imagepulled backoffing
- pending
if it is then it creates the relevant findings (for crashloopbackoff and imagepullbackoff it will by from their code)

**Currently enriches alerts:**
- KubePodNotReady
- KubeContainerWaiting
- KubeStatefulSetReplicasMismatch
- KubeDeploymentReplicasMismatch
- KubernetesDeploymentReplicasMismatch

can be used for KubeJobCompletion, KubeJobNotCompleted to see why the job is stuck but not added yet due to needing to test more to avoid duplicate findings from existing job playbooks

Tested on 
1. Deployments with image pulls
2. Deployments with crashpods
3. 10 different types of pending pods

Future improvements:
Add OOMKill detection as part of its investigation

Screen Shot 2023-04-16 at 13 14 18

Screen Shot 2023-04-16 at 13 15 41

Screen Shot 2023-04-16 at 13 16 51

### CODE REVIEW

**Comment by Avi-Robusta** [comment]
Screen Shot 2023-04-16 at 13 14 18

**Comment by Avi-Robusta** [comment]
Screen Shot 2023-04-16 at 13 19 46

File: src/robusta/core/playbooks/pod_utils/pending_pod_utils.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        event_message = get_unscheduled_message(self.pod)
+        if not event_message:
+            return None
+        # prefer to get message from event if event still exists, it won't exist still for old pods
+        failed_scheduling_events = [
+            pod_event for pod_event in self.pod_events.items if self.is_scheduler_failed_scheduling_event(pod_event)
+        ]
+        if failed_scheduling_events:
+            newest_failed_event = max(failed_scheduling_events, key=lambda x: get_event_timestamp(x))
+            event_message = newest_failed_event[0].message
```
Comment (RoiGlinik) [inline src/robusta/core/playbooks/pod_utils/pending_pod_utils.py:118]: Should it be here [0] ?

Comment (Avi-Robusta) [reply src/robusta/core/playbooks/pod_utils/pending_pod_utils.py:118]: I think your right

File: src/robusta/core/playbooks/pod_utils/imagepull_utils.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        self.namespace = namespace
+
+        self.pod_events: EventList = EventList.listNamespacedEvent(
+            self.namespace, field_selector=f"involvedObject.name={self.pod_name}"
+        ).obj
+
+    def investigate(self, container_status: ContainerStatus) -> Optional[ImagePullOffInvestigation]:
+        for pod_event in self.pod_events.items:
+            error_message = self.get_kubelet_image_pull_error_from_event(pod_event, container_status.image)
+            logging.info(f"for {pod_event} got message: {error_message}")
```
Comment (RoiGlinik) [inline src/robusta/core/playbooks/pod_utils/imagepull_utils.py:163]: some logs here, I assume they were for testing

Comment (Avi-Robusta) [reply src/robusta/core/playbooks/pod_utils/imagepull_utils.py:163]: actually this was the same code  moved from
playbooks/robusta_playbooks/image_pull_backoff_enricher.py

Comment (Avi-Robusta) [reply src/robusta/core/playbooks/pod_utils/imagepull_utils.py:163]: we might not still need it

File: src/robusta/core/model/pods.py [hunk 3]
```diff
@@ -134,6 +134,34 @@ def pod_limits(pod: Pod) -> PodResources:
...
+            requests = container.object_at_path(["resources", resource_attribute.name])  # requests or limits
+            for resource_type in requests.keys():
+                if resource_type in standard_resources:
+                    continue
+                if resource_type not in total_resources:
+                    total_resources[resource_type] = float(requests[resource_type])
+                else:
+                    total_resources[resource_type] += float(requests[resource_type])
+        except Exception:
+            logging.error(f"failed to parce {resource_attribute.name} {container.resources}", exc_info=True)
```
Comment (RoiGlinik) [inline src/robusta/core/model/pods.py:161]: typo

File: playbooks/robusta_playbooks/pod_investigator_enricher.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        pods = [resource]
+    else:
+        # if the kind is Deployment", "DaemonSet", "ReplicaSet", "StatefulSet"
+        selector = build_selector_query(resource.spec.selector)
+        pods = PodList.listNamespacedPod(namespace=resource.metadata.namespace, label_selector=selector).obj.items
+    for pod in pods:
+        pod_issue = detect_pod_issue(pod)
+        if pod_issue == PodIssue.NoneDetected:
+            continue
+        report_pod_issue(event, pods, pod_issue)
```
Comment (RoiGlinik) [inline playbooks/robusta_playbooks/pod_investigator_enricher.py:60]: wrote to you about this , something here looked weird to me

**Review by RoiGlinik — APPROVED** [review]
Looks good . added some comments you can merge after them .

**Pull request merged**